### PR TITLE
Implement movupw parsing

### PR DESCRIPTION
**Pull request by tnachen**
Implement movupw parsing

### CODE REVIEW

File: assembly/src/v1/parsers/stack_ops.rs [hunk 1]
```diff
@@ -43,9 +43,25 @@ pub fn parse_movup(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), As
...
+        "2" => {
+            _span_ops.push(Operation::SwapW);
+            _span_ops.push(Operation::SwapW2);
+        }
+        "3" => {
+            _span_ops.push(Operation::SwapW);
+            _span_ops.push(Operation::SwapW2);
+            _span_ops.push(Operation::SwapW3);
+        }
+        _ => return Err(AssemblyError::extra_param(_op)),
```
Comment (bobbinth) [inline assembly/src/v1/parsers/stack_ops.rs:61]: The error returned form here should be created with `invalid_param()`.

File: assembly/src/v1/parsers/stack_ops.rs [hunk 2]
```diff
@@ -43,9 +43,25 @@ pub fn parse_movup(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), As
     unimplemented!()
 }
 
-/// TODO: implement
 pub fn parse_movupw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+    if _op.num_parts() != 2 {
```
Comment (bobbinth) [inline assembly/src/v1/parsers/stack_ops.rs:47]: We should check here if `op.num_parts()` is less than 2. In such a case, we should return `missing_param()` error. Incases when `op.num_parts()` is greater than two, we should return `extra_param()` (as is done now).

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good. I've added a couple of comment inline, and a couple more general comments:

1. Let's remove leading underscores from all variable names (leading underscores are used for unused variables only).
2. Let's add documentation comments similar to how we do for other implemented parsers.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! A few more comments:

Let's combine the two match statements. A single statement could look as follows:
```Rust
match op.num_parts() {
    0..=1 => return Err(AssemblyError::missing_param(op)),
    2 => match op.parts()[1] {
        ...
    }
    _ => return Err(AssemblyError::extra_param(op)),
}
```

Also, carrying over from the previous review:
* Let's remove leading underscores from all variable names.
* Let's add documentation comments similar to how we do for other implemented parsers.

File: assembly/src/v1/parsers/stack_ops.rs [hunk 3]
```diff
@@ -43,9 +43,26 @@ pub fn parse_movup(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), As
     unimplemented!()
 }
 
-/// TODO: implement
-pub fn parse_movupw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Translates movupw assembly instruction to VM operations.
```
Comment (bobbinth) [inline assembly/src/v1/parsers/stack_ops.rs:46]: I'd prefer a slightly more expanded comment. Maybe something like:

```Rust
/// Translates movupw.x assembly instruction to VM operations.
///
/// Specifically:
/// * movupw.2 is translated into SWAPW SWAPW2
/// * movupw.3 is translated into SWAPW SWAPW2 SWAPW3
```

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks great! I left one minor nit in the comments. Also, could you make sure `rustfmt` tests pass? You can make sure rustfmt and clippy are happy  by running:

```
cargo fix --allow-staged --allow-dirty --all-targets --all-features && cargo fmt && cargo clippy  
```

The above should work on MacOS and Linux.

**Comment by tnachen** [comment]
Ah, seems like it would be nice to have rustfmt run in a pre-commit hook

**Review by bobbinth — APPROVED** [review]
Everything looks great! Thank you!

**Pull request merged**